### ISSUE TITLE
FunctionFailedException.InnerException is always of type System.Exception

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-durable-extension#361]
### Issue
When an activity function or a sub-orchestration throws an exception, it propogates up to the parent orchestrator function as a `FunctionFailedException`. It contains an `InnerException` value which has the same message as the original exception, but the type information is not preserved. Rather, it is always a `System.Exception`.

### Expected
Ideally, the original exception is preserved.

### Root cause
The problem is with the serialization settings between when the exception is observed by the activity function executor (type information is written) and the parent orchestrator (type information is ignored). The fix may require changes to the Durable Task Framework.

**Comment by armadak** [comment]
Any update on this issue?
In my opinion this is vital for proper error handling for activity functions, especially when using CallActivityWithRetryAsync

**Comment by cgillum** [comment]
@armadak Still trying to figure out the right way to fix this. It turns out the serialization of exceptions in .NET is quite tricky for a variety of different reasons.

**Comment by andrew-vandenbrink** [comment]
```
"{\"ExceptionType\":\"Microsoft.Azure.WebJobs.FunctionFailedException\",\"InnerExceptionType\":\"DurableTask.Core.Exceptions.TaskFailedExceptionDeserializationException\"}"
```
TaskFailedExceptionDeserializationException
Instead of trying to serialize/deserialize the whole exception object, 

Can you not just throw the exception where it contains 

```
public class ActivityException : Exception{

public string ActivityName{get;private set;}

public string ExceptionTypeName{get;private set;}
public string ExceptionTypeMessage{get;private set;}

public string InnerExceptionTypeName{get;private set;}
public string InnerExceptionTypeMessage{get;private set;}
}
```

I need to have at least ActivityName or FunctionName where the exception being thrown

**Comment by shibayan** [comment]
@cgillum I tried simple solution. Adding TypeNameHandling to exception serializer settings.
Exception objects always inherit `Exception` class, they are serialized to base class even if they do not exist.
https://github.com/Azure/azure-functions-durable-extension/compare/dev...shibayan:exception-serializer-setting

```csharp
try
{
    await context.CallActivityWithRetryAsync<string>("Function1_Hello", new RetryOptions(TimeSpan.FromSeconds(5), 1)
    {
        Handle = ex =>
        {
            // It's work
            return ex.InnerException is InvalidOperationException;
        }
    }, null);
}
catch (FunctionFailedException ex) when (ex.InnerException is InvalidOperationException)
{
    // It's Work
}
```

![image](https://user-images.githubusercontent.com/1356444/53231848-306a9e00-36cd-11e9-814b-9a37fec2c118.png)

Thanks

**Comment by cgillum** [comment]
@shibayan Interesting - I'm wondering how I could I could have overlooked this 😅. Thank you for sharing the example. We'll take a look and see if we can use this approach to help resolve this item.

**Comment by cgillum** [comment]
Resolved with [v1.8.0](https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.DurableTask/1.8.0) release.
